리뷰 요청

SIKU-KR opened the pull request:
감쟈드립니다~😆

SIKU-KR committed:
기능 구현완료. 테스트 코드 작성 시작.

SIKU-KR committed:
테스트코드 작성완료

SIKU-KR committed:
commit

goldentrash commented:
`main` 함수의 내용이 상당히 복잡하네요. 주석을 통해서 가독성에 신경을 많이 써주신 것 같지만, 함수 자체의 복잡도를 줄여보시는건 어떨까요?

SIKU-KR committed:
리뷰 내용 반영:
메인함수 단순화

SIKU-KR commented:
반영했습니다 !

speculatingwook commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,108 @@
     public static void main(String[] args) {
-        // TODO: 프로그램 구현
+        // 1. 구입 금액 입력
+        Integer amount = null;
```
바로 선언하지 않고 null로 먼저 선언하신 이유가 있나요?

SIKU-KR replied:
아 ! 그러게요. 
이산님 피드백 참고하면서 Extract Method 를 하다가 놓친것 같아요 
수정하겠습니당

speculatingwook commented on src/main/java/lotto/Lotto.java:
```diff
@@ -1,20 +1,56 @@
 
-    // TODO: 추가 기능 구현
+    @Override
+    public String toString() {
```
객체지향의 다형성의 관점에서 이렇게 작성한 코드는 너무 좋다고 생각됩니다!

speculatingwook commented on src/main/java/lotto/Rank.java:
Enum 클래스 잘 작성해주셨습니다!

speculatingwook commented on src/main/java/lotto/User.java:
범식님이 생각하시는 유저의 역할(기능)과 책임은 어떤 것이라고 생각하고 코드를 작성하셨나요?

SIKU-KR replied:
사용자의 Lotto 들을 보관하고 검증하고 마지막 검증단계에서 당첨금계산의 책임이 있다고 생각하고 코드를 작성했습니다.

speculatingwook commented on docs/README.md:
미리 어떻게 코드를 작성하실지 정리해 두는 습관 자체는 너무 좋은 것 같습니다

speculatingwook commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,108 @@
+                System.out.println("구입금액을 입력해 주세요.");
+                int input = Integer.parseInt(Console.readLine());
+                validateAmount(input);
+                amount = input;
```
여기에서 바로 return으로 값을 반환하지 않고 함수의 마지막에 반환하신 이유가 있나요?

SIKU-KR replied:
while 문의 일종의 break; 개념으로 생각을 했는데요. 
솔직히 지금 생각해보니 어떤 방법이 좋은지 잘 모르겠습니다. 
저자리에서 return을 하게되는 것이 좋을까요?

goldentrash replied:
늦게 확인해서 맥락은 확인 안 하고 그냥 적어보지만,,, `early return`이라는 키워드로 검색해보시면 관련 자료가 많을 겁니다!

speculatingwook commented on src/main/java/lotto/Application.java:
범식님이 생각하시는 Application.java의 역할과 기능은 어디까지라고 생각하시나요?

SIKU-KR replied:
초반에 역할을 UI라고 생각을 하고 시작했는데, 코드를 작성하다보니 지키지 못한 부분들이 있네요. 수정해보겠습니다.

speculatingwook commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,108 @@
+        return amount;
+    }
+
+    private static void validateAmount(int amount) {
```
그리고 생각하신 역할과 기능에 금액을 확인하는 것이 Application.java에 포함되어 있나요?

speculatingwook commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,108 @@
+        }
+    }
+
+    private static Lotto makeWinningLotto(String userInputForWinnings) throws NumberFormatException {
```
WinningLotto를 반환하고 있는데, 당첨 로또를 반환하는 것은 범식님이 생각하시는 Application.java의 역할에 포함되어 있나요?

SIKU-KR replied:
Lotto Class의 새로운 생성자로 로직을 옮겼습니다.

speculatingwook replied:
아 그리고 각 클래스들의 역할과 책임이 무엇인지 주석으로 달아주심 감사드리겠습니다!

speculatingwook commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,108 @@
+        return new Lotto(numberSet.stream().sorted().toList());
+    }
+
+    private static void validateLottoNumber(Integer integer){
```
이 로또 번호를 검증하는 로직은 Application.java의 책임이 맞나요?

SIKU-KR replied:
Lotto 클래스의 public static 메소드로 옮겼습니다

speculatingwook commented on src/main/java/lotto/Lotto.java:
검증 로직까지 잘 작성해주셨다고 생각합니다. 다만, 앞서 봤던 Application.java에서 Lotto를 반환하는 로직이 여기에 있는 것이 더 적합하지 않나 생각합니다!

speculatingwook reviewed (commented):
코드를 보면 어느정도 고민을 하거나 잘 짜려고 노력하신 부분들이 많이 보입니다. 제가 달아드린 대부분의 리뷰는 객체지향적인 관점에서 달아드렸습니다.

제가 틀린 부분이 있을 수 있습니다. 그래서 추가적으로 의견이 있으시다면 언제든지 물어봐주심 됩니다. (아마 다시 코드 리팩토링하고 보내주시면  또 다른 리뷰(의견)를 달아드릴 것 같슴다).

speculatingwook commented on src/main/java/lotto/User.java:
```diff
@@ -0,0 +1,80 @@
+        return ret.toString();
+    }
+
+    public String getResultString(Lotto winLotto, Integer bonus) {
```
이 함수 같은 경우에는
1. 결과를 구한다.
2. 문자열로 변환한다.

의 두가지 기능이 들어가서 분리를 시키는 게 더 좋다고 생각합니다! 만약, 비즈니스 로직이 바뀌어서 다른 결과를 보여줘야 하거나, 혹은 결과 자체가 달라지는 경우에도 이 함수를 사용할 수 있는지 한번 생각해보면 좋을 것 같습니다!

speculatingwook commented on src/main/java/lotto/User.java:
```diff
@@ -0,0 +1,80 @@
+        return totalEarnings;
+    }
+
+    private static void getStatisticsString(StringBuilder ret, Map<Rank, Integer> rankCount) {
```
여기 함수들이 전반적으로 멤버 변수가 무엇이고, 출력하는 값이 어떤 값인지 파악하기가 쉽지 않습니다. 함수당 코드를 15줄로 줄이더라도, 각자의 역할이 명확하지 않고 이해하기 어렵다면 긴 코드 못지 않다고 생각합니다!

SIKU-KR committed:
리뷰 내용 반영:
1. Application.java 클래스의 역할을 확실히 했습니다. (UI 기능)
2. 리뷰에서 언급된 validateNumber 함수와 makeWinningLotto의 경우 Lotto.java 에 위치하는 것이 좋다고 생각하여 이동했습니다.
3. 함수 Detail 적인 부분들 수정했습니다.

SIKU-KR commented:
리뷰 내용 반영: 
1. Application.java 클래스의 역할을 확실히 했습니다. (UI 기능)
2. 리뷰에서 언급된 validateNumber 함수와 makeWinningLotto의 경우 Lotto.java 에 위치하는 것이 좋다고 생각하여 이동했습니다.
3. 함수 Detail 적인 부분들 수정했습니다.

User.java 클래스의 내용은 아직 수정하지 못했습니다. 그 위에 리뷰들에 대한 commit입니다!

speculatingwook reviewed (commented):
각 클래스(객체)가 어떤 역할, 책임을 가지고 서로 상호작용하고 있는지 -> 그리고 그 의도대로 작성이 된 코드가 맞는지 한번만 검토 부탁드립니다!(각 클래스 위에 주석으로 정리해주시면 너무 좋을 것 같슴다)

이후 주시는 pull request에 대해 처음부터 다시 리뷰해드리겠습니다!

SIKU-KR committed:
리뷰내용반영

SIKU-KR commented:
아 리뷰하기 너무 복잡할 것 같아 . close 하고 다시열어보겠습니다.
